onnxruntime/core/framework/sequential_executor.cc

```diff
@@ -71,32 +71,37 @@ Status SequentialExecutor::Execute(const SessionState& session_state, const std:
 
     // sync before compute
     int queue_id = p_op_kernel->KernelDef().ExecQueueId();
-    for (int input_index = 0; input_index < op_kernel_context.InputCount(); ++input_index) {
-      Fence_t fence = op_kernel_context.InputFence(input_index);
-      if (fence) {
-        auto execution_provider_type = p_op_kernel->Node().GetExecutionProviderType();
-        if (OrtMemTypeCPUInput == p_op_kernel->KernelDef().InputMemoryType(input_index)) {
-          execution_provider_type = kCpuExecutionProvider;
+    
+    // Fence is created only if queue_id != 0. Retrieving fence with input_index is costly,
+    // so we can do fence check only when queue_id != 0.
+    if (queue_id != 0) {
```

Comment:
>queue_id != 0 [](start = 8, length = 13)

I think it may not be safe. For example, think of a case where there's a MemCpy node copying data asynchronously from CPU to GPU, and then there's a CUDA kernel using that buffer. The CUDA kernel would be executed on queue_id==0, but it still need to wait for the fence until the previous copy to GPU finished.

IMO A better fix is to detect the need of copy in execution planning and set a bool on kernel to indicate if sync is needed.

Reply:
Good point, will fix it.